ext/nokogiri_ext_xmlsec/nokogiri_sign.c

```diff
@@ -144,23 +157,30 @@ VALUE sign_with_certificate(VALUE self, VALUE rb_opts) {
     goto done;
   }
   
-  // load certificate and add to the key
-  if(xmlSecCryptoAppKeyCertLoadMemory(dsigCtx->signKey,
-                                      (xmlSecByte *)certificate,
-                                      certificateLength,
-                                      xmlSecKeyDataFormatPem) < 0) {
-    rb_exception_result = rb_eSigningError;
-    exception_message = "failed to load certificate";
-    goto done;
+  if(keyName) {
+    // set key name
+    if(xmlSecKeySetName(dsigCtx->signKey, (xmlSecByte *)keyName) < 0) {
+      rb_exception_result = rb_eSigningError;
+      exception_message = "failed to set key name";
+      goto done;
+    }
   }
 
-  // set key name
-  if(xmlSecKeySetName(dsigCtx->signKey, (xmlSecByte *)keyName) < 0) {
-    rb_exception_result = rb_eSigningError;
-    exception_message = "failed to set key name";
-    goto done;
+  if(certificate) {
+    // load certificate and add to the key
+    if(xmlSecCryptoAppKeyCertLoadMemory(dsigCtx->signKey,
+                                        (xmlSecByte *)certificate,
+                                        certificateLength,
+                                        xmlSecKeyDataFormatPem) < 0) {
+      rb_exception_result = rb_eSigningError;
+      exception_message = "failed to load certificate";
+      goto done;
+    }
   }
 
+  // add <dsig:Signature/> node to the doc
```

Comment:
I moved it so that if any of the other methods failed, it would leave the document unmodified. And now I see that signNode will get leaked if there are any intervening errors. man, I've been doing ruby too long, and haven't done manual memory management. I'll fix that tomorrow.